feat(storage): suppport disk object store

wenym1 opened the pull request:
## What's changed and what's your intention?

In this PR, we implement a disk object store for future spill to disk support. The local file system is used to maintain the path hierarchy and store the object data.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
part of #2384

skyzh commented:
Do we have plan to support multiple object store backends for Hummock? If we want to spill to disk, I guess we are running two backends. I'm also thinking whether disk should be an object store. Maybe we only need to have disk interface for spill-to-disk, without making it a kind of object store>

wenym1 commented:
> Do we have plan to support multiple object store backends for Hummock? If we want to spill to disk, I guess we are running two backends. I'm also thinking whether disk should be an object store. 

Yes, in my development plan, we will have two object store, one for local and one for remote. The sstable table store will hold the two object store and do the routing according to the highest bit of the sstable id. A more progressed PR is #2384, and since it's too large, I split in into several smaller PR like this one. 

> Maybe we only need to have disk interface for spill-to-disk, without making it a kind of object store?

Since the spilled files will be in SST format, making it an object store can share some logic when we read SST and make the code neater. In future PR I will introduce a stream-like uploader for object store so that we don't have to buffer the whole SST in memory when we are using disk object store.

skyzh commented:
> we will have two object store
Looks good!

skyzh commented on src/storage/src/store_impl.rs:
```diff
@@ -117,6 +121,15 @@ impl StateStoreImpl {
                     in_mem_object_store.set_compactor_shutdown_sender(shutdown_sender);
+                } else if let ObjectStoreImpl::Disk(disk_object_store) = object_store.as_ref() {
+                    tracing::info!("start a compactor for local disk object store");
+                    let (_, shutdown_sender) = Compactor::start_compactor(
```
Shall we limit the capability of this compactor? e.g., only allow it to compact files on local disk.

wenym1 replied:
The compactor is unaware of where the file is coming from. It just fetches table id from meta and data from object store. The code here is only for probably using disk object store as the remote ground true object, similar to using in memory object store.

wenym1 commented on src/storage/src/object/disk.rs:
```diff
@@ -0,0 +1,431 @@
+    use crate::object::{ObjectError, ObjectResult};
+
+    pub async fn ensure_file_dir_exists(path: &Path) -> ObjectResult<()> {
+        if let Some(dir) = path.parent() {
```
I think it's not necessary. If the current path is a folder, in the subsequent open it will fail.

wenym1 commented on src/storage/src/object/disk.rs:
```diff
@@ -0,0 +1,431 @@
+        Ok(())
+    }
+
+    async fn read(&self, path: &str, block_loc: Option<BlockLocation>) -> ObjectResult<Bytes> {
```
Do you mean after we finish using a file handle we recycle it in a cache instead of closing it?

skyzh replied:
Yes. Also we should use a single file handler (file object) for a file, and then avoid using mutex around them. If there are multiple requests to a file, we should use `pread` to do positioned read.

Little-Wallace replied:
+1. 
But it seems that tokio does not support pread.

wenym1 commented on src/storage/src/object/disk.rs:
```diff
@@ -0,0 +1,431 @@
+    use std::fs::Metadata;
+    use std::path::Path;
+
+    use tokio::fs::{create_dir_all, File, OpenOptions};
```
Just added a simple bench. The bench runs serially in a single thread. The bench result are as followed. Each r/w is operating on 1MB data.
```
fs_operation/tokio      time:   [124.42 ms 126.53 ms 128.76 ms]
tokio stat: Write 1310 times, avg 5.948854961832061ms. Read 1310 times, avg 4.315267175572519ms

fs_operation/std        time:   [27.044 ms 27.805 ms 28.629 ms]
std stat: Write 3270 times, avg 0.0027522935779816515ms. Read 3270 times, avg 0.046788990825688076ms
```

Little-Wallace replied:
> > Do we have plan to support multiple object store backends for Hummock? If we want to spill to disk, I guess we are running two backends. I'm also thinking whether disk should be an object store.
> 
> Yes, in my development plan, we will have two object store, one for local and one for remote. The sstable table store will hold the two object store and do the routing according to the highest bit of the sstable id. A more progressed PR is #2384, and since it's too large, I split in into several smaller PR like this one.
> 
> > Maybe we only need to have disk interface for spill-to-disk, without making it a kind of object store?
> 
> Since the spilled files will be in SST format, making it an object store can share some logic when we read SST and make the code neater. In future PR I will introduce a stream-like uploader for object store so that we don't have to buffer the whole SST in memory when we are using disk object store.

Would we implement it as a hybrid-storage?  Just like following:

```rust
pub struct HybridObjectStore {
    local: Box<dyn ObjectStore>,
    remote: Box<dyn ObjectStore>,
}

impl ObjectStore for HybridObjectStore {
}
```

wenym1 replied:
I have thought about it. If we implement it in this way, in the current `ObjectStore` interface, I can't find an elegant way to specify which object store to use. In the future, such `HybridObjectStore` can be used with cache semantic, i.e. we first lookup the local object store, which acts as a local cache, and on cache miss, we then lookup the remote object store.

BugenZhao commented on src/storage/src/hummock/cache.rs:
```diff
@@ -933,9 +934,12 @@ mod tests {
-            assert!(!(*old_entry).is_in_lru());
             assert!((*new_entry).is_in_cache());
-            assert!((*new_entry).is_in_lru());
+            #[cfg(debug_assertions)]
+            {
+                assert!(!(*old_entry).is_in_lru());
+                assert!((*new_entry).is_in_lru());
+            }
```
Use `debug_assert` here?

wenym1 replied:
`is_in_lru` only compiles when debug_assertions is enabled, while code generated from `debug_assert` is `if cfg!(debug_asserts){...}`, whose body is still compiled even when `debug_assertions` is disabled. Therefore, if we use `debug_asserts` here we will get a compile error.

TennyZhuang replied:
ditto

BugenZhao commented on src/storage/src/object/disk.rs:
```diff
@@ -0,0 +1,431 @@
+                ))
+            })?;
+
+            if read_size != buf.len() {
```
It seems `read_exact` returning `Ok` already implies this?

BugenZhao commented on src/storage/src/object/disk.rs:
```diff
@@ -0,0 +1,431 @@
+        if let Some(dir) = path.parent() {
+            // `create_dir_all` will not error even if the directory already exists.
+            create_dir_all(dir).await.map_err(|e| {
+                ObjectError::internal(format!(
```
How about adding a new `ObjectError` variant like `Disk` and impl `From<io::Error>` for it?

wenym1 commented:
Just added an opened file cache. Used `spawn_blocking` as a work-around to call `file.read_exact_at`.

Also extracted the common logic of deduplicating concurrent request on cache entry in `cache.lookup_with_request_dedup`. Block cache and meta cache, which uses the `cache.lookup_for_request`, are refactored accordingly.

TennyZhuang commented on src/storage/src/hummock/cache.rs:
```diff
@@ -924,6 +958,7 @@ mod tests {
 
             cache.release(new_entry);
             assert!((*new_entry).is_in_cache());
+            #[cfg(debug_assertions)]
             assert!((*new_entry).is_in_lru());
```
```suggestion
            debug_assert!((*new_entry).is_in_lru());
```

I guess this will work

wenym1 replied:
It's not working if we are running `cargo bench`, where unit test will be compile while `debug_assertions` is not enabled.

The code of `is_in_lru` only exists when `debug_assertions` is enabled.
```
#[cfg(debug_assertions)]
fn is_in_lru(&self) -> bool {
    (self.flags & IN_LRU) > 0
}
```
, while the generated code of `debug_assert` is 
```
if cfg!(debug_assertions) {
 ...
}
```
, and usage for `is_in_lru` still exists even if `debug_assertions` is not enabled, and when running `cargo bench`, the compilation will faill.

skyzh replied:
okay, I originally expected that `debug_asserts` is equivalent to `#[cfg(debug_assertions)]`.

Little-Wallace commented on src/storage/src/object/disk.rs:
```diff
@@ -0,0 +1,495 @@
+        T: Send + 'static,
+        F: FnOnce() -> ObjectResult<T> + Send + 'static,
+    {
+        spawn_blocking(f).await.map_err(|e| {
```
It seems that `asyncify` will block one physical thread of tokio. 
How about create some independent threads to serve IO request?

wenym1 replied:
`spawn_blocking` is separated from `spawn` and is dedicated to run tokio tasks that can probably be blocked in place. It will use threads separated from threads that run normal tokio tasks(async/await tasks).

Reference: 
https://docs.rs/tokio/0.2.25/tokio/task/fn.spawn_blocking.html
https://docs.rs/tokio/0.2.25/tokio/index.html#cpu-bound-tasks-and-blocking-code

skyzh commented on src/storage/src/hummock/cache.rs:
```diff
@@ -677,6 +680,37 @@ impl<K: LruKey, T: LruValue> LruCache<K, T> {
+        F: FnOnce() -> VC,
+        VC: Future<Output = HummockResult<(T, usize)>>,
+    {
+        match self.lookup_for_request(hash, key.clone()) {
```
Is it possible for `self.lookup_for_request` to take `&K` instead of `K`?

wenym1 replied:
No. `lookup_for_request` includes potential insert to a map, and this needs the ownership of `K`.

skyzh commented on src/storage/src/hummock/cache.rs:
```diff
@@ -677,6 +680,37 @@ impl<K: LruKey, T: LruValue> LruCache<K, T> {
+        match self.lookup_for_request(hash, key.clone()) {
+            LookupResult::Cached(entry) => Ok(entry),
+            LookupResult::WaitPendingRequest(recv) => {
+                let entry = recv.await.map_err(HummockError::other)?;
```
Better to return the original error to the caller, instead of `channel closed` or something unclear. Here we can only get channel closed error.

skyzh replied:
Previously, `moka` will return `Arc<Error>` to all callers.

Little-Wallace replied:
The original error may not support `Clone`?

skyzh commented on src/storage/src/object/disk.rs:
```diff
@@ -0,0 +1,495 @@
+        }
+    }
+
+    async fn readv(&self, path: &str, block_locs: Vec<BlockLocation>) -> ObjectResult<Vec<Bytes>> {
```
`block_locs: impl AsRef<[BlockLocation]>`, so that users can either pass a reference `&[BlockLocation { xxx }]` or a vector.

skyzh commented on src/storage/src/object/mem.rs:
```diff
@@ -103,6 +103,14 @@ impl InMemObjectStore {
+impl Drop for InMemObjectStore {
+    fn drop(&mut self) {
+        if let Some(sender) = self.compactor_shutdown_sender.lock().take() {
+            let _ = sender.send(());
```
better to join the compactor thread (future), otherwise there will be some unexpected errors on SIGINT.

Little-Wallace replied:
Why notify compactor shutdown in objectstore? 
We shall notify compactor shutdown in state-store closed.

skyzh reviewed (approved):
Rest LGTM, good work!

Little-Wallace reviewed (approved):
LGTM
